### DISCUSSION TITLE
Security for organizations each with services and users

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [fastapi-users/fastapi-users#1188] [Q&A] [answered]
Assume a dockerized FastApi service (call it Zippy) which is available from a cloud platform.

Organizations register for Zippy with login credentials.

An organization creates one or more Zippy-based services.

Users access a Zippy service anonymously or login.

Need FastApi to handle the security access for organizations and its individual services ie.

```
Cloud
    Zippy - a docker'ized FastApi-based service 
        Org-1
            zippy-service-11
            zippy-service-12
            ...
            zippy-service-1n
        Org-2
            zippy-service-21
            zippy-service-22
            ...
            zippy-service-2p
        ...
        Org-m
            zippy-service-m1
            zippy-service-m2
            ...
            zippy-service-mr
```
Is this possible with FastApi-Users?

If so, are there any suggestions or caveats? Thanks

**Comment by frankie567** [comment]
FastAPI Users help you to implement user registration and authentication. From what I understand, you need to handle *Organization* objects.

Typically, you'll need to define those `Organization` in your models and link them to one or several `User` that have access to this organization.

**Comment by dineshbvadhia** [comment]
Agree about models for Organizations and Users and linking them.

Wrt security with jwt: An Organization has one or more Services, and each Service has its own Users. 

Is it correct to say: The set of organizations need its own jwt 'secret'? And, each Service needs its own 'secret' too for its Users? Or, is it one 'secret' for all organizations and the users of services?

**Reply by frankie567** [reply]
Not necessarily IMO. If your secret is strong and kept safe on your server, I don't see anything wrong with using the same one for the whole app.

**Reply by dineshbvadhia** [reply]
I just want to make sure we are on the same page: Each Organization+Service+Users combination (see initial post above) is a docker fastapi-based service which runs on it own server.

If a strong secret is kept safe on the server (inside the docker container?), it can be used by every Organization+Service+Users instance.

Is this correct?

**Reply by frankie567** [reply, marked as answer]
Well, the best way to set this kind of values is to set it through environment variables. You should avoid to hardcode it in the source code.

This way, you can build your Docker image once, and when executing it set the value for your secret and other settings.
